File: python/KPA.py
from Rotor96Crypto import Rotor96Crypto


# Load passwords from file
def load_passwords(filename):
    with open(filename, 'r') as file:
        return [line.strip() for line in file if line.strip()]


def load_text(filename):
    with open(filename, 'r') as file:
        return file.read().strip()


def KPA_dictionary_attack(ciphertext, known_plaintext, password_file):
    passwords = load_passwords(password_file)
    possible_decipher = {}
    for password in passwords:
        try:
            # Rotor96Crypto.set_key(password)
            decrypted_text = Rotor96Crypto.encdec(
                Rotor96Crypto.DEC, password, ciphertext)
            if decrypted_text[:len(known_plaintext)] == known_plaintext:
                print(f"Key found: {password}")
                print(f"Decrypted text: {decrypted_text}")
                possible_decipher[password] = decrypted_text
                
        except Exception as e:
            print(f"Error with key {password}: {e}")
    
    if len(possible_decipher) == 0:
        print("Key not found in dictionary.")
        return None
    
    return possible_decipher


if __name__ == "__main__":

    ciphertext = load_text("../known_data/ciphertext1.txt")
    known_plaintext = load_text("../known_data/known_plaintext.txt")
    password_filename = "../passwords"  # File containing passwords

    print("Deciphering...")
    possible_decipher = KPA_dictionary_attack(ciphertext, known_plaintext, password_filename)
    print("Done.")
    
    if len(possible_decipher) > 0:
        with open(f"decrypted_text/decrypted_text1.txt", 'w') as f:
            f.write("key \t decrypted_text\n")
            for password, decrypted_text in possible_decipher.items():
                f.write(f"{password} \t {decrypted_text}\n")

File: rotor96Crypto/KPA.java
// Known Plaintext Attach
package rotor96Crypto;

// import Rotor96Crypto;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class KPA {

    // Load passwords from file
    public static List<String> loadPasswords(String filename) throws IOException {
        List<String> passwords = new ArrayList<>();
        BufferedReader reader = new BufferedReader(new FileReader(filename));
        String line;
        while ((line = reader.readLine()) != null) {
            if (!line.trim().isEmpty()) {
                passwords.add(line.trim());
            }
        }
        reader.close();
        return passwords;
    }

    // Load text from file
    public static String loadText(String filename) throws IOException {
        BufferedReader reader = new BufferedReader(new FileReader(filename));
        StringBuilder textBuilder = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            textBuilder.append(line).append("\n");
        }
        reader.close();
        return textBuilder.toString().trim();
    }

    // KPA dictionary attack
    public static Map<String, String> knownPlaintextAttack(String ciphertext, String knownPlaintext, String passwordFile) {
        Map<String, String> possibleDecipher = new HashMap<>();
        try {
            List<String> passwords = loadPasswords(passwordFile);
            for (String password : passwords) {
                try {
                    // Assuming Rotor96Crypto.set_key and Rotor96Crypto.encdec are methods in a class
                    String decryptedText = Rotor96Crypto.encdec(Rotor96Crypto.DEC, password, ciphertext);
                    if (decryptedText.startsWith(knownPlaintext)) {
                        System.out.println("Key found: " + password);
                        System.out.println("Decrypted text: " + decryptedText);
                        possibleDecipher.put(password, decryptedText);
                    }
                } catch (Exception e) {
                    System.err.format("Error with key %s: %s%n", password, e.getMessage());
                    e.printStackTrace();
                }
            }
        } catch (IOException e) {
            System.err.format("Error reading passwords from file: %s%n", e.getMessage());
            e.printStackTrace();
        }

        if (possibleDecipher.isEmpty()) {
            System.out.println("Key not found in dictionary.");
            return null;
        }

        return possibleDecipher;
    }

    public static void main(String[] args) {
        // Example usage
        try {
            String ciphertext = loadText("known_data/ciphertext1.txt");
            String knownPlaintext = loadText("known_data/known_plaintext.txt");
            String passwordFile = "known_data/passwords";

            System.out.println("Deciphering..");
            Map<String, String> result = knownPlaintextAttack(ciphertext, knownPlaintext, passwordFile);
            System.out.println("Done.");
            
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
